app/screens/create_or_edit_channel/index.tsx

```diff
@@ -0,0 +1,280 @@
+        }
+
+        if (goBack) {
+            popTopScreen();
+        } else {
+            dismissModal();
+        }
+    };
+
+    const [appState, dispatch] = useReducer((state: RequestState, action: RequestAction) => {
+        switch (action.type) {
+            case RequestActions.START:
+                emitCanSaveChannel(true);
+                Keyboard.dismiss();
+                return {
+                    ...state,
+                    error: '',
+                    saving: true,
+                };
+            case RequestActions.COMPLETE:
+                emitCanSaveChannel(false);
+                close(true);
+
+                // TODO: setChannelDisplayName
+                return {
+                    ...state,
+                    error: '',
+                    saving: false,
+                };
+            case RequestActions.FAILURE:
+                emitCanSaveChannel(false);
+                return {
+                    ...state,
+                    error: action.error,
+                    saving: false,
+                };
+
+            default:
+                return {
+                    ...state,
+                };
+        }
+    }, {
+        error: '',
+        saving: false,
+        type: channel?.type as ChannelType || General.OPEN_CHANNEL,
+    });
+
+    useEffect(() => {
+        const button = {
+            testID: 'edit_channel.save.button',
+            id: EDIT_CHANNEL_ID,
+            enabled: false,
+            showAsAction: 'always',
+            color: theme.sidebarHeaderTextColor,
+            text: formatMessage({id: 'mobile.edit_channel', defaultMessage: 'Save'}),
+        };
+
+        if (!channel) {
+            button.id = CREATE_CHANNEL_ID;
+            button.text = formatMessage({id: 'mobile.create_channel', defaultMessage: 'Create'});
+        }
+
+        setRightButton(button);
+        setButtons(componentId, {
+            leftButtons: [],
+            rightButtons: [{...button, enabled: false}] as never[],
+        });
+    }, []);
+
+    useEffect(() => {
+        const update = Navigation.events().registerComponentListener({
+            navigationButtonPressed: ({buttonId}: {buttonId: string}) => {
+                switch (buttonId) {
+                    case CLOSE_CHANNEL_ID:
+                        close(true);
+                        break;
+                    case CREATE_CHANNEL_ID:
+                        onCreateChannel();
+                        break;
+                    case EDIT_CHANNEL_ID:
+                        onUpdateChannel();
+                        break;
+                }
+            },
+        }, componentId);
+
+        return () => {
+            update.remove();
+        };
+    }, [displayName, header, purpose]);
+
+    const onTypeChange = (typeText: ChannelType) => {
+        setType(typeText);
+    };
+
+    // if a channel was provided, we are editing a channel
+    const editing = Boolean(channel);
```

Comment:
0/5 Personal preference: I like these derived constants to be at the top of the component, just after the `useX` oneliners and before effects, callbacks and reducers.

Reply:
done.